subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -282,6 +283,43 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec {
         executed(":producer", ":consumer")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/16061")
+    def "does not detect missing dependencies even for complicated filters"() {
+        buildFile """
+            task prepareBuild {
+                outputs.file("app/foo.txt")
+                doLast {}
+            }
+
+            def sources = fileTree("app") {
+                include("**/*.txt")
+                exclude("**/*generated*")
+                builtBy(prepareBuild)
+            }
+
+            task generateA {
```

Comment:
Just a comment on the readability of the test. It would have helped if instead of `generateA`, the task would have been called `consumesResultOfPrepareBuildAndGeneratesAInSameDirectory` or something like that.